### ISSUE TITLE
Spinwait performance regression

### ISSUE THREAD

**Issue by hcedwar** [kokkos/kokkos#935]
performance testing with `kokkos/benchmarks/policy_performance`
```
./policy_performance.host 1000 1000 32 20 10 10 2 1 1 112  # static case
./policy_performance.host 1000 1000 32 20 10 10 2 1 2 112  # dynamic case
```
Exposes spinwait performance problem on POWER8.
Merge PR #910 introduced the problem.

**Comment by ndellingwood** [comment]
@dsunder @hcedwar 
On KNL there seems to be no regression for this case, possible slight improvement for static case (~3%)

**Comment by mhoemmen** [comment]
@jjellio this looks familiar perhaps, though the PR in question only got merged into develop 10 days ago.

**Comment by ndellingwood** [comment]
@dsunder @hcedwar 
Huge regression with KNL after this commit:

hwloc-bind core:0-63 ./policy_performance.host 1000 1 32 25000 1 1 8 1 1 100
Regression ~490% (not a typo)

**Comment by ndellingwood** [comment]
I checked out the commit prior to #937, build, and reran and the times looked as previous, may want to revert that commit before the snapshot.

**Comment by hcedwar** [comment]
On the master branch POWER architecture does not have assembly instructions to pause so the spinwait immediately called 'sched_yield()' instead of flooding memory loads, preserve this behavior for POWER while using assembly to pause on other platforms.

**Comment by hcedwar** [comment]
For POWER architecture a spinwait within team collective operations gives best performance when it retries the wait a large number of iterations before yielding.  However, a spinwait for global collective operations gives best performance when it immediately yields.

**Comment by ndellingwood** [comment]
@hcedwar @crtrott Merged #971; as discussed the performance tests of concern on Power will be tracked in a separate issue; integration testing should be ready to resume.

**Comment by nmhamster** [comment]
@hcedwar @crtrott @dsunder @ndellingwood - please see PR #973 for POWER8 spin-wait yields.
